### PR TITLE
Fix sendAcrossNetwork for multi-dimensional Views

### PR DESCRIPTION
**Pull request by masterleinad**
This should fix the failing tests in https://github.com/ORNL-CEES/DataTransferKit/pull/574 while not impacting performance for the one-dimensional case.

### CODE REVIEW

**Comment by Rombur** [comment]
When using this PR inside DTK, I get
```
/home/dtk/packages/Search/src/details/ArborX_DetailsUtils.hpp(125): error: no instance of overloaded function "Kokkos::deep_copy" matches the argument list
            argument types are: (const Kokkos::CudaUVMSpace::memory_space, Kokkos::View<int **, Kokkos::LayoutRight, Kokkos::CudaUVMSpace>, const Kokkos::View<int **, DeviceTypeKokkos_Compat_KokkosCudaWrapperNode>)
          detected during:
            instantiation of "auto ArborX::Details::create_layout_right_mirror_view_and_copy(const ExecutionSpace &, const View &, std::enable_if<<expression>, void>::type *) [with View=Kokkos::View<int **, DeviceTypeKokkos_Compat_Kokkos
CudaWrapperNode>, ExecutionSpace=Kokkos::CudaUVMSpace::memory_space]"
/home/dtk/packages/Search/src/details/ArborX_DetailsDistributor.hpp(283): here
            instantiation of "void ArborX::Details::Distributor<DeviceType>::doPostsAndWaits(const ExecutionSpace &, const ExportView &, size_t, const ImportView &) const [with DeviceType=DeviceTypeKokkos_Compat_KokkosCudaWrapperNode, Ex
ecutionSpace=Kokkos::CudaUVMSpace::execution_space, ExportView=Kokkos::View<int **, DeviceTypeKokkos_Compat_KokkosCudaWrapperNode>, ImportView=Kokkos::View<int *, Kokkos::CudaUVMSpace::memory_space, Kokkos::MemoryTraits<1U>>]"
/home/dtk/packages/Search/src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp(215): here
            instantiation of "std::enable_if<Kokkos::is_view<View>::value, void>::type ArborX::Details::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(const ExecutionSpace &, const ArborX::Details::Distributor<DeviceType> &, Vi
ew, View::non_const_type) [with DeviceType=DeviceTypeKokkos_Compat_KokkosCudaWrapperNode, ExecutionSpace=Kokkos::CudaUVMSpace::execution_space, View=Kokkos::View<int **, DeviceTypeKokkos_Compat_KokkosCudaWrapperNode>]"
/home/dtk/packages/Meshfree/test/tstDetailsCommunicationHelpers.cpp(69): here
            instantiation of "void Helper<DeviceType>::checkSendAcrossNetwork(MPI_Comm, const View1 &, const View2 &, const View2 &, __nv_bool &, Teuchos::FancyOStream &) [with DeviceType=DeviceTypeKokkos_Compat_KokkosCudaWrapperNode, Vi
ew1=Kokkos::View<int *, DeviceTypeKokkos_Compat_KokkosCudaWrapperNode>, View2=Kokkos::View<int **, DeviceTypeKokkos_Compat_KokkosCudaWrapperNode>]"
/home/dtk/packages/Meshfree/test/tstDetailsCommunicationHelpers.cpp(126): here
            instantiation of "void DetailsDistributedSearchTreeImpl_send_across_network_UnitTest<DeviceType>::runUnitTestImpl(Teuchos::FancyOStream &, __nv_bool &) const [with DeviceType=DeviceTypeKokkos_Compat_KokkosCudaWrapperNode]"
/home/dtk/packages/Meshfree/test/tstDetailsCommunicationHelpers.cpp(241): here
```

**Comment by masterleinad** [comment]
Thanks for checking! I haven't looked at `CUDA` yet.

**Comment by aprokop** [comment]
@masterleinad Can you please describe what the problem is, and how you address it?

**Comment by masterleinad** [comment]
Sure. The issue is that we need `LayoutRight` for MPI communication in its current form both for the send buffer and the receive buffer. This requires changing the layout at least when we are using CUDA but also to be careful when applying the permutation.
The idea is to change the layout as late as possible, i.e. after applying the permutation, for the data to send so that we have to deal with the type-punned data as little as possible.
There are not that many actual changes (mostly https://github.com/arborx/ArborX/pull/353/commits/e26474fa63e1106edc5783fc29ec52f820395b67).

It seems that we need another intermediate copy when changing the layout and copying between different memory spaces.

**Comment by masterleinad** [comment]
One thing that can likely be improved is to have a better separation between `doPostsAndWaits` and `sendAcrossNetwork` and possibly factor applying the permutation in `doPostsAndWaits` out. I would rather do it in a follow-up, though, to fix `DTK`.

File: src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp [hunk 1]
```diff
@@ -256,28 +190,27 @@ DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
...
                Kokkos::MemoryTraits<Kokkos::Unmanaged>>
-      import_buffer(imports.data(), imports.size());
+      import_buffer(imports_layout_right.data(), imports_layout_right.size());
 
-  distributor.doPostsAndWaits(space, export_buffer, num_packets, import_buffer);
-#endif
+  distributor.doPostsAndWaits(space, exports, num_packets, import_buffer);
+
+  auto tmp_view = Kokkos::create_mirror_view_and_copy(
+      typename View::device_type{}, imports_layout_right);
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp:212]:
```suggestion
      space, imports_layout_right);
```

File: src/details/ArborX_DetailsUtils.hpp [hunk 2]
```diff
@@ -394,6 +514,20 @@ template <typename SrcViewType, typename DstViewType>
 #endif
 // clang-format on
 
+template <typename View>
+View create_empty_view(std::string const &label)
```
Comment (dalg24) [inline src/details/ArborX_DetailsUtils.hpp:518]: Move to `Details::` namespace

File: src/details/ArborX_DetailsUtils.hpp [hunk 3]
```diff
@@ -394,6 +514,20 @@ template <typename SrcViewType, typename DstViewType>
...
+{
+  constexpr int rank = View::rank_dynamic;
+  return View(label, rank > 0 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+              rank > 1 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+              rank > 2 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+              rank > 3 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+              rank > 4 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+              rank > 5 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+              rank > 6 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+              rank > 7 ? 0 : KOKKOS_IMPL_CTOR_DEFAULT_ARG);
```
Comment (dalg24) [inline src/details/ArborX_DetailsUtils.hpp:528]: I suspect `View("label", View::array_layout())` does what you want

File: test/tstDistributedSearchTree.cpp [hunk 4]
```diff
@@ -622,3 +622,76 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(boost_comparison, DeviceType, ARBORX_DEVICE_TYPES)
...
+
+  // send 1 packet to rank k
+  // receive comm_size packets
+  Kokkos::View<int **, DeviceType> u_exp("u_exp", comm_size, DIM);
+  Kokkos::parallel_for(Kokkos::RangePolicy<ExecutionSpace>(0, comm_size),
+                       KOKKOS_LAMBDA(int i) {
+                         for (int j = 0; j < DIM; ++j)
+                           u_exp(i, j) = i + j * comm_rank;
+                       });
+  Kokkos::fence();
```
Comment (dalg24) [inline test/tstDistributedSearchTree.cpp:683]:
```suggestion
```

File: test/tstDistributedSearchTree.cpp [hunk 5]
```diff
@@ -622,3 +622,76 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(boost_comparison, DeviceType, ARBORX_DEVICE_TYPES)
...
+  Kokkos::View<int *, DeviceType> ranks_u("", comm_size);
+  ArborX::iota(ExecutionSpace{}, ranks_u, 0);
+
+  Kokkos::View<int **, DeviceType> u_ref("u_ref", comm_size, DIM);
+  Kokkos::parallel_for(Kokkos::RangePolicy<ExecutionSpace>(0, comm_size),
+                       KOKKOS_LAMBDA(int i) {
+                         for (int j = 0; j < DIM; ++j)
+                           u_ref(i, j) = comm_rank + i * j;
+                       });
+  Kokkos::fence();
```
Comment (dalg24) [inline test/tstDistributedSearchTree.cpp:694]:
```suggestion
```

File: test/tstDistributedSearchTree.cpp [hunk 6]
```diff
@@ -622,3 +622,76 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(boost_comparison, DeviceType, ARBORX_DEVICE_TYPES)
...
+
+    // NOTE here we assume that the reference solution is sized properly
+    auto v_imp = Kokkos::create_mirror(typename View2::memory_space(), v_ref);
+
+    ArborX::Details::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+        typename DeviceType::execution_space{}, distributor, v_exp, v_imp);
+
+    // FIXME not sure why I need that guy but I do get a bus error when it
+    // is not here...
+    Kokkos::fence();
```
Comment (dalg24) [inline test/tstDistributedSearchTree.cpp:644]: Has anybody looked into this?

File: src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp (hunk 1 above)
Comment (dalg24) [reply src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp:212]: I am wrong.  please do `View::execution_space` and comment

**Comment by masterleinad** [comment]
@dalg24 I addressed your commits and the tests are passing.

**Pull request merged**